tools/nni_cmd/config_schema.py

```diff
@@ -25,6 +28,71 @@ def setPathCheck(key):
     '''check if path exist'''
     return And(os.path.exists, error=SCHEMA_PATH_ERROR % key)
 
+class AlgoSchema:
+    def __init__(self, algo_type):
+        assert algo_type in ['tuner', 'assessor', 'advisor']
+        self.algo_type = algo_type
+        self.algo_schema = {
+            Optional('codeDir'): setPathCheck('codeDir'),
+            Optional('classFileName'): setType('classFileName', str),
+            Optional('className'): setType('className', str),
+            Optional('classArgs'): dict,
+            Optional('includeIntermediateResults'): setType('includeIntermediateResults', bool),
+            Optional('gpuIndices'): Or(int, And(str, lambda x: len([int(i) for i in x.split(',')]) > 0), error='gpuIndex format error!'),
+        }
+        self.builtin_keys = {
+            'tuner': 'builtinTunerName',
+            'assessor': 'builtinAssessorName',
+            'advisor': 'builtinAdvisorName'
+        }
+        self.builtin_name_schema = {}
+        for k, n in self.builtin_keys.items():
+            self.builtin_name_schema[k] = {Optional(n): setChoice(n, *get_all_builtin_names(k+'s'))}
+
+        self.customized_keys = set(['codeDir', 'classFileName', 'className'])
+
+    def validate_class_args(self, class_args, algo_type, builtin_name):
+        if not builtin_name or not class_args:
+            return
+        meta = get_builtin_algo_meta(algo_type+'s', builtin_name)
+        if meta and 'accept_class_args' in meta and meta['accept_class_args'] == False:
+            raise SchemaError('classArgs is not allowed.')
+
+        validator = create_validator_instance(algo_type+'s', builtin_name)
+        if validator:
+            try:
+                validator.validate_class_args(**class_args)
+            except Exception as e:
+                raise SchemaError(str(e))
+
+    def missing_customized_keys(self, data):
+        return self.customized_keys - set(data.keys())
+
+    def validate_extras(self, data, algo_type):
```

Comment:
better to add docstring, it is not clear what is the meaning of `data`

Reply:
`data` is customized keys?

Reply:
it includes keys within `classArgs`?

Reply:
Thanks, docstring added for AlgoSchema.
The `schema` and `data` naming follows the python schema package convention, the `schema` is a group of rules, `data` is the data to be checked whether it follow the rules: 
https://pypi.org/project/schema/

`AlgoSchema('tuner')` creates the schema of `tuner` section of experiment config file.
`AlgoSchema('assessor')` creates the schema of `assessor` section of experiment config file.
`NNIConfigSchema()` creates the schema of entire experiment config file